File: Advance-Algorithms-and-Complexity/Programming-Assignment-4/circuit_design/circuit_design.py
# python3
# bugged solution 
# not working correctly
n, m = map(int, input().split())
clauses = [ list(map(int, input().split())) for i in range(m) ]
st = [set()]

# This solution tries all possible 2^n variable assignments.
# It is too slow to pass the problem.
# Implement a more efficient algorithm here.
def dfs1(adj, visited, x, order):
    for nxt in adj[x]:
        if not visited[nxt]:
            visited[nxt] = True
            dfs1(adj, visited, nxt, order)
            order.append(nxt)


def dfs2(adj, visited, x, stt):
    stt.add(x)
    for nxt in adj[x]:
        if not visited[nxt]:
            visited[nxt] = True
            dfs2(adj, visited, nxt, stt)
    


def implicationGraph(adj, adj2):
    for clause in clauses:
        x, y, nx, ny = [0] * 4
        if clause[0] > 0:
            x = clause[0] - 1
            nx = x + n
        else:
            x = - clause[0] - 1 + n
            nx = x - n
        if clause[1] > 0:
            y = clause[1] - 1
            ny = y + n
        else:
            y = - clause[1] - 1 + n
            ny = y - n
        
        adj[nx].append(y)
        adj[ny].append(x)
        adj2[x].append(ny)
        adj2[y].append(nx)
    


def settingUp():
    adj = [[] for _ in range(2 * n)]
    adj2 = [[] for _ in range(2 * n)]
    implicationGraph(adj, adj2)

    order = []
    visited = [False for _ in range(2 * n)]

    for i in range(2 * n):
        if not visited[i]:
            visited[i] = True
            dfs1(adj2, visited, i, order)
            order.append(i)
    order.reverse()

    for i in range(len(adj)):
        visited[i] = False
    
    for x in order:
        if not visited[x]:
            stt = set()
            visited[x] = True
            dfs2(adj, visited, x, stt)
            st.append(stt)  


def isSatisfiable():
    settingUp()
    for stt in st:
        for x in stt:
            if stt.__contains__(x + n) or stt.__contains__(x - n):
                return None
    
    result = [ - 1 for _ in range(n)]
    for stt in st:
        for i in stt:
            x, ineg = 0, False
            if i >= n:
                x = i - n
                ineg = True
            else:
                x = i
                ineg = False
            if result[x] == - 1:
                if ineg:
                    result[x] = 1
                else:
                    result[x] = 0
            
    return result


result = isSatisfiable()
if result is None:
    print("UNSATISFIABLE")
else:
    print("SATISFIABLE")
    print(" ".join(str(-i if result[i - 1] == 1 else i) for i in range(1, n + 1)))

File: Advance-Algorithms-and-Complexity/Programming-Assignment-4/circuit_design/circuit_design.cpp
// complete solution
#include <bits/stdc++.h>
using namespace std;

struct Clause {
    int firstVar;
    int secondVar;
};

struct TwoSatisfiability {
    int numVars;
    vector<Clause> clauses;
    vector<set<int> > sccs;

    TwoSatisfiability(int n, int m) :
        numVars(n),
        clauses(m)
    {  }
    

    void constructImplicationGraph(vector<vector<int> > &adj, vector<vector<int> > &adjR)
    {
        //vector<Clause> ::iterator it=clauses.begin(); it!=clauses.end(); ++it
        for (int i = 0; i < clauses.size(); ++i)
        {   Clause clause = clauses[i];
            int x, y, notX, notY;

            if (clause.firstVar > 0)
            {
                x = clause.firstVar - 1;
                notX = x + numVars;
            }
            else
            {
                x = -clause.firstVar - 1 + numVars;
                notX = x - numVars;
            }

            if (clause.secondVar > 0)
            {
                y = clause.secondVar - 1;
                notY = y + numVars;
            }
            else
            {
                y = -clause.secondVar - 1 + numVars;
                notY = y - numVars;
            }

            adj[notX].push_back(y);
            adj[notY].push_back(x);
            adjR[y].push_back(notX);
            adjR[x].push_back(notY);
        }
    }

    void dfs1(vector<vector<int>> &adj, bool *visited, int x, vector<int> &order)
    {
        for (int i=0; i<adj[x].size(); ++i)
        {   int nextX = adj[x][i];
            if (!visited[nextX])
            {
                visited[nextX] = true;
                dfs1(adj, visited, nextX, order);
                order.push_back(nextX);
            }
        }
    }

    void dfs2(vector<vector<int> > &adj, bool *visited, int x, set<int> &scc)
    {
        scc.insert(x);
        for (int i = 0; i < adj[x].size() ; ++i)
        {   int nextX = adj[x][i];
            if (!visited[nextX])
            {
                visited[nextX] = true;
                dfs2(adj, visited, nextX, scc);
            }
        }
    }

bool isSatisfiable(vector<int> &result)
{   set<int> ::iterator it;
    for (int i = 0; i<sccs.size(); ++i)
    {   set<int> scc = sccs[i];
        for (it = scc.begin();it!=scc.end(); ++it)
        {   int x = *it;
            if (scc.find(x + numVars) != scc.end() || scc.find(x - numVars) != scc.end())
                return false;
        }

    }

    for (int i = 0; i < result.size(); i++)
        result[i] = -1;
    for (int i = 0; i < sccs.size(); ++i)
    {
        set<int> scc = sccs[i];
        for (it = scc.begin(); it != scc.end(); ++it)
        {
            int i = *it;
            int x;
            bool isNegative;
            if (i >= numVars)
            {
                x = i - numVars;
                isNegative = true;
            }
            else
            {
                x = i;
                isNegative = false;
            }
            if (result[x] == -1)
            {
                if (isNegative)
                    result[x] = 1;
                else
                    result[x] = 0;
            }
        }
    }
    return true;
}

void setup()
{
    vector<vector<int> > adj(2 * numVars), adjR(2 * numVars);
    constructImplicationGraph(adj, adjR);
    vector<int> order ;
    bool *visited = new bool[2 * numVars];
    for(int i = 0; i < 2 * numVars; ++i)
        visited[i] = false;
    for (int i = 0; i < 2 * numVars; i++)
    {
        if (!visited[i])
        {
            visited[i] = true;
            dfs1(adjR, visited, i, order);
            order.push_back(i);
        }
    }
    reverse(order.begin(),order.end());
    for (int i = 0; i < adj.size(); i++)
        visited[i] = false;

    for (int i = 0; i < order.size(); ++i)
    {   int x = order[i];
        if (!visited[x])
        {
            set<int> scc;
            visited[x] = true;
            dfs2(adj, visited, x, scc);
            sccs.push_back(scc);
        }
    }
}
};
void run()
{
    int n, m;
    cin >>n >> m;

    TwoSatisfiability twoSat(n, m);
    int u, v;
    for (int i = 0; i < m; ++i)
    {   cin >> u >> v;
        twoSat.clauses[i].firstVar = u;
        twoSat.clauses[i].secondVar = v;
    }

    twoSat.setup();

    vector<int> result(n);
    if (twoSat.isSatisfiable(result))
    {
        cout << "SATISFIABLE\n";
        for (int i = 1; i <= n; ++i)
        {
            if (result[i - 1] == 1)
            {
                cout << -i; 
            }
            else
            {
                cout << i;
            }
            if (i < n)
            {
                cout << " ";
            }
            else
            {
                cout << "\n";
            }
        }
    }
    else
    {
        cout << "UNSATISFIABLE\n";
    }
}

int main() {
    ios::sync_with_stdio(false);
    run();
    return 0;
}
